### DISCUSSION TITLE
Inconsistent Confidence Scores with Different Batch Sizes in MultiModalPredictor

### DISCUSSION THREAD

**Discussion by Call-Paul** [autogluon/autogluon#4638] [Q&A]
**Description**: I've been experiencing an issue with MultiModalPredictor in AutoGluon where the confidence scores for image predictions vary depending on the batch size. When I predict a single image in isolation, I get a different confidence score than when I predict the same image as part of a batch with multiple images.

Setup:

Environment: Dockerized FastAPI application
Model Type: Image classification with MultiModalPredictor
Inference Setup: Using model.predict_proba() for confidence scores
Problem Details: When using model.predict_proba():

A single image, when predicted in a batch size of 1, yields a certain confidence score.
The same image, when predicted as part of a larger batch (e.g., batch size of 4), results in a different confidence score.
This behavior occurs despite efforts to set deterministic settings, including:

Setting torch.manual_seed() and torch.cuda.manual_seed().
Enabling torch.no_grad() to disable gradients during inference.
Ensuring model.eval() mode on any underlying PyTorch models (although I understand AutoGluon should handle this automatically).
Reproducible Example (Simplified):

```
from autogluon.multimodal import MultiModalPredictor
import pandas as pd
import torch

# Load model
model_path = '/app/model/my_model'
predictor = MultiModalPredictor.load(model_path)

# Set seeds for determinism
torch.manual_seed(3)
torch.cuda.manual_seed(3)

# Single image prediction
image_data = ... # Load or define your image data here
single_prediction = predictor.predict_proba(pd.DataFrame({'image': [image_data]}))

# Batch prediction (same image repeated)
batch_prediction = predictor.predict_proba(pd.DataFrame({'image': [image_data, image_data, image_data, image_data]}))

print("Single image confidence:", single_prediction)
print("Batch image confidence:", batch_prediction)
```
**Troubleshooting Attempts:**

Verified that torch.no_grad() is active during inference to prevent any dropout or unwanted randomness.
Confirmed that MultiModalPredictor should handle eval() mode internally for batch normalization and dropout layers.
Tried forcing the evaluation mode manually on any accessible underlying model layers.
Logged the prediction times, confidence scores, and checked for differences between single and batch predictions.
Expected Behavior: I expected that predicting the same image with different batch sizes would yield identical confidence scores. Typically, model inference modes should ensure consistent output irrespective of batch size.

**Questions**:

Has anyone else encountered this issue with MultiModalPredictor in AutoGluon?
Could batch normalization or other layers be interacting differently in AutoGluon’s internal model based on batch size?
Are there any recommended fixes or best practices to ensure consistent confidence scores for the same image across different batch sizes?
Additional Notes: I'm wondering if there's some internal batching behavior in MultiModalPredictor or in its handling of batch normalization that could explain these inconsistencies. Any insights or suggestions would be greatly appreciated!

Thank you in advance for any help or suggestions!
